File: src/main/java/com/github/yuppieflu/stats/rest/TransactionsController.java
package com.github.yuppieflu.stats.rest;

import com.github.yuppieflu.stats.rest.dto.Statistic;
import com.github.yuppieflu.stats.rest.dto.Transaction;
import com.github.yuppieflu.stats.service.StorageService;
import com.github.yuppieflu.stats.service.domain.Measurement;
import com.github.yuppieflu.stats.service.domain.Status;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
public class TransactionsController {

    private final StorageService storageService;

    @PostMapping("/transactions")
    public ResponseEntity<Void> transactions(@RequestBody Transaction transaction) {
        Status status = storageService.addMeasurement(new Measurement(transaction));
        switch (status) {
            case REJECTED: return ResponseEntity.noContent().build();
            case PROCESSED: return ResponseEntity.status(HttpStatus.CREATED).build();
            default: throw new IllegalStateException("Unknown status [" + status + "]");
        }
    }

    @GetMapping("/statistics")
    public Statistic statistics() {
        return Statistic.from(storageService.getStatistic());
    }
}

File: src/main/java/com/github/yuppieflu/stats/rest/TransactionsController.kt
package com.github.yuppieflu.stats.rest

import com.github.yuppieflu.stats.rest.dto.Statistic
import com.github.yuppieflu.stats.rest.dto.Transaction
import com.github.yuppieflu.stats.rest.ex.FutureTransactionException
import com.github.yuppieflu.stats.service.StorageService
import com.github.yuppieflu.stats.service.domain.Measurement
import com.github.yuppieflu.stats.service.domain.Status
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RestController

@RestController
class TransactionsController(private val storageService: StorageService) {

    @PostMapping("/transactions")
    fun transactions(@RequestBody transaction: Transaction): ResponseEntity<Void> {
        val status = storageService.addMeasurement(Measurement(transaction))
        return when (status) {
            Status.REJECTED -> ResponseEntity.noContent().build()
            Status.PROCESSED -> ResponseEntity.status(HttpStatus.CREATED).build()
            Status.INVALID -> throw FutureTransactionException()
        }
    }

    @GetMapping("/statistics")
    fun statistics(): Statistic {
        return Statistic(storageService.getStatistic())
    }
}